Command API – take two

Reinmar opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: The command API has been redesigned. The `Command` methods are now public and consistent. Commands can be used in a standalone mode (without the editor). The `CommandCollection` was introduced and replaced the `Map` of commands used in `editor.commands`. Closes ckeditor/ckeditor5#2917.

  Besides changes mentioned in this point and in the "Breaking changes" section, other minor changes were done:

  * `Editor#execute()` now accepts multiple command arguments.
  * `Command#value` property was standardized.

BREAKING CHANGES: The `Command`'s protected `_doExecute()` and `_checkEnabled()` methods have been replaced by public `execute()` and `refresh()` methods.

BREAKING CHANGES: The `Command`'s `refreshState` event was removed and one should use `change:isEnabled` in order to control and override its state.

BREAKING CHANGES: The `core/command/command` module has been moved to the root directory (so the `Command` class is `core/command~Command` now).

BREAKING CHANGES: The `Command#refresh()` method is now automatically called on `editor.document#changesDone`.

BREAKING CHANGES: The `editor.commands` map was replaced by a `CommandCollection` instance so `editor.commands.set()` calls need to be replaced with `editor.commands.add()`.

---

### Additional information

Requires https://github.com/ckeditor/ckeditor5-utils/pull/163.

I left `ToggleAttributeCommand` in `src/command` because I'll be refactoring this bit in a separate ticket (to not make this too big).

Reinmar committed:
First batch of changes – CommandCollection and new Command class.

Reinmar committed:
Tests.

Reinmar committed:
Fixed API docs.

Reinmar commented:
Ouch! While working on refactoring the `EnterCommand` I noticed it does this:

```js
	execute() {
		const doc = this.editor.document;
		const batch = doc.batch();

		doc.enqueueChanges( () => {
			enterBlock( this.editor.data, batch, doc.selection, doc.schema );

			this.fire( 'afterExecute', { batch } );
		} );
	}
```

I recall now why we didn't know how to implement automatic `afterExecute` firing in the past :D. But I guess I'll find some more or less straightforward solution.

Reinmar commented:
While refactoring all the commands I realised that:

* `isEnabled` should be set to `true` by default because there are number of commands which are always on. This can be done by implementing default `refresh()` in `Command` which will also reduce code duplication.
* `isEnabled` should be checked in `execute`'s listener automatically. Every command requires it and but it's easy to forget about this. Especially in commands which are always `true` (when you look at their code only), but which can be blocked from outside (about which you may forget).
* I was also considering whether every command should be automatically refreshed by its constructor. That might be useful... but it's not possible to do it automatically in `Command#constructor()` because at that point some properties of a child command classes will not be set yet (you can't access `this` before calling `super()`). So perhaps this would be nice but it can't be done automatically anyway.

Reinmar committed:
Provided Command with default implementation of refresh() and changed default values of command properties to better match how commands usually look.

Reinmar committed:
Implemented automatic command execution blocking and got rid of unnecesary CommandInterface.

Reinmar commented:
I added the mentioned changes updated the suggested commit message.

Reinmar committed:
Added default implementation of execute() because the constructor depends on that.

oleq commented:
> isEnabled should be set to true by default because there are number of commands which are always on.

Can you explain this thing a little bit? Because at first glance it feels so much like [this pitfall](http://dev.ckeditor.com/ticket/10439) in v4 and it really scares me ;-) 

I suppose you meant something like "all commands' states should be refreshed upon `editor#ready`" or something like this?

Reinmar commented:
> Can you explain this thing a little bit?

We're talking here about the default implementation of `refresh()`. Each command can implement its own and most of the commands do that. However, a couple of simple commands are just always enabled. And hence, to avoid unnecessary code duplication and potential issues, I implemented base `Command#refresh()` method which unconditionally sets `isEnabled` to `true`. I also changed the default `isEnabled` set by the constructor.

However, this is only the "base" code and initial state. Every command is refreshed on `changesDone` anyway, so once data is loaded, commands change their state according to their own rules or base implementation if it fits.

There are two issues, though:

* Sometimes `changesDone` may not be called (in special cases where editor starts really empty). We know that and I think it's tracked in some issue, but it's a hypothetical situation now, so let's skip it.
* But you made me realise that it'd be safer if the initial value of `isEnabled` set by the constructor was `false`. This would mean that before first `refresh()` is called you won't be able to execute any of the commands. This would make perfect sense so I'll change it.

oleq commented:
> And hence, to avoid unnecessary code duplication and potential issues, I implemented base Command#refresh() method which unconditionally sets isEnabled to true. I also changed the default isEnabled set by the constructor.

This is exactly what I wanted to know – that the thing is not hard coded but controlled by a generic refresh. Thanks!

pjasiun commented on src/commandcollection.js:
```diff
@@ -0,0 +1,103 @@
+	 *
+	 * @param {String} commandName The name of the command.
+	 */
+	execute( commandName, ...args ) {
```
Do we get much having the shortcut:
```js
editor.commands.execute( 'bold', { forceState: true } );
```
for:
```js
editor.commands.get( 'bold' ).execute( { forceState: true } );
```

I mean, the second one is consistent with:
```js
editor.commands.get( 'bold' ).isEnabled;
editor.commands.get( 'bold' ).value;
```
If we go with the first one shouldn't we have shortcuts for the rest?

pjasiun replied:
I mean, I like `editor.commands.get( 'bold' ).execute( { forceState: true } );` because it's clear here that `{ forceState: true }` is just a parameter of the `execute` method. Nothing more, nothing less.

pjasiun replied:
But the question "Do we get much having the shortcut" is a real, not a rhetorical question. I also have a feeling that we may need such shortcut.

Reinmar replied:
I implemented `CommandCollection#execute()` to keep there the error. That was meant to reduce `Editor` size. The real shortcut for the external user is `editor.execute( 'bold', { forceState: true } )`. Having `Editor#execute` and `CommandCollection#execute` was only meant to shorten the `Editor` class and make it potentially easier to reimplement from scratch.

I hope that's fine.

pjasiun replied:
Now I see, that there is `editor.execute` that make sense to keep `commands.execute` then.

Reinmar committed:
Changed the default value of isEnabled to false. It's set to true on first refresh().

Reinmar commented:
Back on review.

Reinmar commented:
I also pushed all the other commands refactored in https://github.com/ckeditor/ckeditor5/tree/t/ckeditor5-core/88.

Its build fails (https://travis-ci.org/ckeditor/ckeditor5/builds/242366798?utm_source=github_status&utm_medium=notification) due to a bug in API docs validation (https://github.com/ckeditor/ckeditor5-dev/issues/229) and can be ignored for now. The tests and everything pass.

pjasiun commented:
I checked these changes and its R+ from me.

Reinmar commented:
When merging t/ckeditor5-core/88 branches to master branches of respective repositories please use this merge commit:

```
Other: Aligned the implementation to the new Command API (see https://github.com/ckeditor/ckeditor5-core/issues/88).

BREAKING CHANGES: The command API has been changed.
```

scofalik commented on src/command.js:
```diff
@@ -0,0 +1,122 @@
+			if ( !this.isEnabled ) {
+				evt.stop();
+			}
+		}, { priority: 'high' } );
```
Why not `highest` priority?

pjasiun replied:
I can imagine that one want to capture the `execute` event even if the command is disabled in some special cases.

scofalik commented on src/command/toggleattributecommand.js:
```diff
@@ -30,43 +31,29 @@ export default class ToggleAttributeCommand extends Command {
 		 */
 		this.attributeKey = attributeKey;
 
 		/**
```
Is this description still true? It's a bit convoluted and it hides the fact that we use document selection. Maybe we should write it directly that the value is set depending on document's selection.

Reinmar replied:
Hm... I'll be still working on the `ToggleAttributeCommand` so I'll see if I can put this differently.

oskarwrobel commented on src/commandcollection.js:
```diff
@@ -0,0 +1,103 @@
+	 * @param {String} commandName The name of the command.
+	 * @param {module:core/command~Command} command
+	 */
+	add( commandName, command ) {
```
`MarkerCollection` uses `set` method in a similar case. https://github.com/ckeditor/ckeditor5-engine/blob/master/src/model/markercollection.js#L85.

Reinmar replied:
I think it depends on a language used in a certain domain. I think that you're more likely to say "add a command to the editor" rather than "set a command". But I'd say "set a keystroke".

The are many uses of `set()` and `add()` across our code. I don't think we'll be able to justify certain picks :D

Reinmar replied:
One more thing – in `MarkerCollection` there's `set()` because it allows to override values (it even fires `remove` and `add`). We don't expect that in `CommandCollection` for instance – it should throw (it doesn't now, but that was the assumed semantic).
